Pagination component

tylersticka opened the pull request:
## Overview

Adds a Pagination component, similar to [the live version](https://cloudfour-patterns.netlify.app/patterns/components/pagination.html) but with improvements to accessibility, layout and hit target sizes.

## Screenshots

![Screenshot_2020-05-08 Storybook](https://user-images.githubusercontent.com/69633/81438608-e929cb00-9121-11ea-9290-1f148bb46cf1.png)

<img width="680" alt="Screen Shot 2020-05-08 at 11 46 14 AM" src="https://user-images.githubusercontent.com/69633/81438640-f2b33300-9121-11ea-822c-318afe5e512f.png">

## Testing

[**View Component Deploy Preview →**](https://deploy-preview-692--cloudfour-patterns.netlify.app/?path=/docs/components-pagination--example)

- Read the documentation, try links to make sure they work and are correct.
- With a mouse…
  - Verify hover states for all links and direction affordances work.
  - Verify that active state (mouse down) for direction affordances work.
- With a keyboard in Chrome or Safari…
  - Verify that focus states on numbers are visible (but not on mouse interactions).
- From [the canvas view](https://deploy-preview-692--cloudfour-patterns.netlify.app/?path=/story/components-pagination--example)…
  - Toggle dark mode and light mode from the top toolbar.
  - Toggle different breakpoints from the top toolbar.
  - View the "Accessibility" tab and make sure tests pass…
    - Without a theme
    - With dark theme
  - View the "Knobs" tab, focus the "current" field and tap up and down to test that the current number changes and the next and previous numbers change when expected.

(Note that the "knobs" are for visualization purposes and do not incorporate any sort of data validation, so it's possible to break the component by putting in nonsense values.)

---

- Resolves #478

tylersticka commented:
> I noticed in Chrome at smaller viewports the previous and next actions appear to stutter or jump slightly with hover and mouseout.

I tried a few fixes that had absolutely no effect. Then I tried it [outside of Storybook's iframe](https://deploy-preview-692--cloudfour-patterns.netlify.app/iframe.html?id=components-pagination--example) and it doesn't seem to have the same issue.

So I think I'm chalking this up to Chrome handling animation in iframes worse than other browsers.

> I'm having issues with the `visibility-hidden` link in the docs.

Good catch! I've updated this link.

derekshirk commented on src/components/pagination/pagination.scss:
```diff
@@ -0,0 +1,265 @@
+ */
+
+.c-pagination__action.is-previous::before {
+  background-image: svg-load(
+    'icons/arrow-left.svg',
+    color=colors.$primary-brand
+  );
```
❤️

spaceninja commented on src/components/pagination/pagination.scss:
```diff
@@ -0,0 +1,265 @@
+ */
+
+.c-pagination__action.is-previous:hover::before,
+.c-pagination__action.is-next:hover::after {
```
This is another situation where I think Sass nesting could make this more readable:

```scss
.c-pagination__action {
  // link styles

  [aria-current] {
    // current styles
  }

  &.is-previous::before,
  &.is-next::after {
    // icon styles

    &:hover {
      // hover styles
    }

    &:active {
      // active styles
    }
  }
}
```

…etc

tylersticka replied:
But I don't want the states to happen _only_ on the affordances, I want it to happen when you hover over the action element.

tylersticka replied:
(I tried to explain this reasoning in my comment block.)

spaceninja replied:
Oh, I get it now. Sorry, I got mixed up on what the states applied to. Your code is correct. :)

spaceninja commented on src/components/pagination/pagination.scss:
```diff
@@ -0,0 +1,265 @@
+ * pagination dot, as seen in many app interfaces.
+ */
+
+.c-pagination__action[aria-current] .c-pagination__number::after {
```
I find the nested Sass parent selector easier to read for situations like this, because it helps me understand that I'm still dealing with the same element, just in a new context. eg:

```scss
.c-pagination__number::after {
  // indicator styles

  .c-pagination__action[aria-current] & {
    // current state modifier
  }
}
```

tylersticka replied:
- [x] Moved selector for consistency with previous `:not` selector

spaceninja commented on src/components/pagination/demo/mock-pagination-data.ts:
```diff
@@ -0,0 +1,54 @@
+  };
+}
+
+export default function mockPaginationData({
```
Could you add JSDoc-style comment blocks here? It makes reading JS files a lot easier when functions have a clear name + description comment block.

Also explaining the options would be helpful. I've read the function and I still have no idea what `midSize` is.

tylersticka replied:
I've tended toward being pretty lean with demonstrative or prototype JavaScript since it's not intended for production.

I had hoped that linking to two pages that describe what `mid_size` is at the top of the file would suffice.

spaceninja replied:
Ah, I didn't understand that this isn't production JS. That makes more sense, then.

tylersticka replied:
- [x] Added JSDoc-style comments for this file

spaceninja commented on src/components/pagination/demo/mock-pagination-data.ts:
```diff
@@ -0,0 +1,54 @@
+  current = 1,
+  midSize = 2,
+  total = 36,
+} = {}) {
```
I'm not familiar with this options syntax. Are you passing an options object and then setting the default to an empty object?

tylersticka replied:
See: https://simonsmith.io/destructuring-objects-as-function-parameters-in-es6#allowing-the-configuration-object-to-be-optional

spaceninja replied:
Oh cool! #TIL

tylersticka replied:
- [x] Added `@param` comments to make structure more clear

spaceninja commented on src/components/pagination/demo/mock-pagination-data.ts:
```diff
@@ -0,0 +1,54 @@
+  let end = current + midSize;
+
+  if (start < 1) {
+    end += 1 - start;
```
A comment explaining what you're doing here would be helpful

tylersticka replied:
- [x] Added line-by-line comments to this function

spaceninja commented on src/components/pagination/demo/mock-pagination-data.ts:
```diff
@@ -0,0 +1,54 @@
+    start -= end - total;
+  }
+
+  start = Math.max(1, start);
```
I _think_ what you're doing here is saying "take the largest value of either `1` or the `start` value", but a comment would be helpful. Perhaps I'm overly sensitive to this as a newer JS dev, but it's helpful to comment your code with the assumption that a junior developer with less familiarity with the codebase will end up maintaining this code.

I've never been frustrated to find _too many_ comments, but I'm often frustrated by the lack of them.

tylersticka replied:
- [x] Added line-by-line comments to this function

Paul-Hebert commented on src/components/pagination/pagination.scss:
```diff
@@ -0,0 +1,265 @@
+.c-pagination__action.is-next:active::after {
+  filter: brightness(80%);
+  transform: scale(0.95);
+}
```
Nit: I wonder if it would be useful to house these interaction state values as design tokens if they're shared between multiple components?

tylersticka replied:
- [x] Created design tokens for these values and referenced them across multiple components

Paul-Hebert commented on src/components/pagination/pagination.scss:
```diff
@@ -0,0 +1,265 @@
+ *    smaller than the "previous" and "next" affordances.
+ * 2. Fixes an issue where Chrome renders the digits a pixel or two higher
+ *    than in other browsers.
+ * 3. The aforementioned means of positioning the indicator pseudo element.
```
Nit: It might be nice to include more context here. I know what you're alluding to since I've been reading through this file, but if I was a future dev troubleshooting an issue I'd need to dig through this whole file to figure it out

tylersticka replied:
So, when I say "the aforementioned means," I'm talking about the preceding paragraph, which says:

> The inner number element gives us a means of positioning indicator pseudo elements relative to the action content without disrupting the "previous" and "next" affordances.

Do you have suggestions for what additional clarity I could provide?

tylersticka commented on src/components/pagination/pagination.scss:
```diff
@@ -0,0 +1,265 @@
+ * "previous" and "next" affordances.
+ */
+
+.c-pagination__items::before,
```
Because `c-pagination__item` doesn't need to be nested, and it felt weird nesting this block and then _not_ nesting the next (or repeating the `flex: 100%` in two different places).

Paul-Hebert commented on src/components/pagination/pagination.stories.mdx:
```diff
@@ -0,0 +1,61 @@
+- An `aria-current` attribute (typically with a value of `page`) to identify the current page action.
+- `is-previous` or `is-next` classes to identify the _previous_ or _next_ page action.
+
+## Accessibility
```
Yeah these are great. It's awesome to see the care put into making this accessible

tylersticka commented:
@spaceninja @Paul-Hebert I believe I've addressed all of your feedback except those I've left as unresolved conversations (which I've left responses to).
